### DISCUSSION TITLE
Composer versioning uses stability modifiers incorrectly

### DISCUSSION THREAD

**Discussion by Xerkus** [renovatebot/renovate#24989] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Renovate incorrectly assumes stability flags to be equal to stability suffixes in the version, eg `1.2.3@RC1` to be equal to `1.2.3-rc1` 

Part of version normalization:
https://github.com/renovatebot/renovate/blob/6f4c38990400f2caea83172616d08027ede1f30f/lib/modules/versioning/composer/index.ts#L46-L62

And as seen in tests:
https://github.com/renovatebot/renovate/blob/0da50b51604ec918afbd5981ba8c834e03366484/lib/modules/versioning/composer/index.spec.ts#L28-L38

There are three things wrong here:
- stability flags are constraints and never a part of the actual version. Constraint `1.2.3@RC` is a range, equivalent to `>=1.2.3-rc <1.2.4`. Tag datasource with tag `1.2.3@RC` will likely produce invalid result if I read the renovate code right.
- in the presence of stability suffix in the version, unless said suffix is `stable`, composer entirely disregards stability flag. Constraint `1.2.3-alpha@RC` is equivalent to `1.2.3@alpha`
- stability flag can be one of: `stable`, `RC`, `alpha`, `beta` or `dev` with no numerical suffixes allowed. `1.2.3@RC1` is invalid and composer won't parse it.

I discovered this while looking for other non-canonical formats that are supported by composer but not renovate. Eg `1.2.3RC1` or `1.2.3a1`. I would like to provide implementation for same version normalization as used by composer and fix this bug as well.

No reproduction needed:
```
composer require php:~8.2.1@rc1

In VersionParser.php line 521:
                                                                                      
  Could not parse version constraint ~8.2.0@rc1: Invalid version string "~8.2.0@rc1"
```

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment, marked as answer]
Thank you for your analysis of the problem. I have converted this Discussion to a bug report issue directly: #24993

Once you comment there I can assign you that issue, and I will close this one to avoid duplication.

**Comment by Xerkus** [comment]
As an aside, how does it work out getting bug reports as discussions at first? Never seen this approach before.

**Reply by rarkins** [reply]
It's worked out well so far. Although we experimented with using labels (e.g. default to "triage" priority and a "needs requirements" label) we still ended up with our issues polluted with low-quality topics. I _don't_ want to use a stale bot or close too aggressively, so using Discussions - which can stay open forever - seems a nicer balance. Probably no more than 10-20% of "report a problem" turn out to be an actual bug (unlike this one..).

Also unlike this one, sometimes feature requests or bugs take a long conversation until we arrive to the _actual_ thing which needs changing. In such cases it's much better to start over with a clean issue than to have to scroll through pages of history /misunderstandings.
